### PR TITLE
Std module docs improvements

### LINKED ISSUE #92777: Docs for primitive type helper modules should stay in their lane
Summary of primitive types with corresponding std modules:

| Primitive type | `std` module | `std` module docs first line |
| --- | --- | --- |
| [array] | [std::array] | Helper functions and types for fixed-length arrays.
| [char] | [std::char] | A character type.
| [slice] | [std::slice] | A dynamically-sized view into a contiguous sequence, [T].
| [str] | [std::str] | Unicode string slices.
| [pointer] | [std::ptr] | Manually manage memory through raw pointers.
| [f32] | [std::f32] | Constants specific to the f32 single-precision floating point type.
| [f64] | [std::f64] | Constants specific to the f64 double-precision floating point type.

Some of the std modules' documentation sound like documentation for the primitive type itself. Others do not. The `std::slice` and `slice` docs almost look like they are supposed to be identical but there are some minor (perhaps accidental) differences. `std::ptr` has some unique content that may be better fit for the `pointer` docs.

In general, I think the std docs should be very brief and primarily work as a signpost pointing to the primitive type docs. But of course they should also document the module itself, without sounding like "I am char". Some more specific conventions could be established.

[array]: https://doc.rust-lang.org/std/primitive.array.html
[char]: https://doc.rust-lang.org/std/primitive.char.html
[slice]: https://doc.rust-lang.org/std/primitive.slice.html
[str]: https://doc.rust-lang.org/std/primitive.str.html
[pointer]: https://doc.rust-lang.org/std/primitive.pointer.html
[f32]: https://doc.rust-lang.org/std/primitive.f32.html
[f64]: https://doc.rust-lang.org/std/primitive.f64.html
[std::array]: https://doc.rust-lang.org/std/array/index.html
[std::char]: https://doc.rust-lang.org/std/char/index.html
[std::slice]: https://doc.rust-lang.org/std/slice/index.html
[std::str]: https://doc.rust-lang.org/std/str/index.html
[std::ptr]: https://doc.rust-lang.org/std/ptr/index.html
[std::f32]: https://doc.rust-lang.org/std/f32/index.html
[std::f64]: https://doc.rust-lang.org/std/f64/index.html

### PR DESCRIPTION
**Pull request by camsteffen**
My primary goal is to create a cleaner separation between primitive types and primitive type helper modules (fixes #92777). I also changed a few header lines in other top-level std modules (seen at https://doc.rust-lang.org/std/) for consistency.

Some conventions used/established:

 * "The \`Box\<T>` type for heap allocation." - if a module mainly provides a single type, name it and summarize its purpose in the module header
 * "Utilities for the _ primitive type." - this wording is used for the header of helper modules
 * Documentation for primitive types themselves are removed from helper modules
 * provided-by-core functionality of primitive types is documented in the primitive type instead of the helper module (such as the "Iteration" section in the slice docs)

I wonder if some content in `std::ptr` should be in `pointer` but I did not address this.

### CODE REVIEW

**Comment by m-ou-se** [comment]
It makes me super happy to see someone working on documentation consistency. I'll try to make the time to review this soon.

File: library/core/src/primitive_docs.rs [hunk 1]
```diff
@@ -746,19 +793,22 @@ mod prim_slice {}
...
-/// let hello = "Hello, world!";
-///
-/// // with an explicit type annotation
-/// let hello: &'static str = "Hello, world!";
+/// let hello_world = "Hello, World!";
 /// ```
 ///
-/// They are `'static` because they're stored directly in the final binary, and
-/// so will be valid for the `'static` duration.
+/// Here we have declared a string literal, also known as a string slice.
```
Comment (m-ou-se) [inline library/core/src/primitive_docs.rs:804]: While every string literal is a string slice, not every string slice is a literal, so let's avoid implying they are the same thing. (A 'string slice' could also refer to (a part of) a `String`, for example.)

File: library/core/src/primitive_docs.rs [hunk 2]
```diff
@@ -731,12 +731,59 @@ mod prim_array {}
...
+/// [`.splitn`]: slice::splitn
+/// [`.chunks`]: slice::chunks
+/// [`.windows`]: slice::windows
 #[stable(feature = "rust1", since = "1.0.0")]
 mod prim_slice {}
 
 #[doc(primitive = "str")]
 //
-/// String slices.
+/// Unicode string slices.
```
Comment (m-ou-se) [inline library/core/src/primitive_docs.rs:786]: I'd be careful with using the term 'unicode string'. That's used in some other languages/on other platforms to refer to `[u32]` or `[u16]` string representations. You could mention UTF-8 specifically, but that's already mentioned below and might be a bit too much detail for the first line.

File: library/core/src/any.rs [hunk 3]
```diff
@@ -1,5 +1,5 @@
-//! This module implements the `Any` trait, which enables dynamic typing
-//! of any `'static` type through runtime reflection.
+//! The `Any` trait enables dynamic typing of any `'static` type through runtime
+//! reflection.
```
Comment (m-ou-se) [inline library/core/src/any.rs:2]: This module also provides `TypeId` and `type_name`. While related to `Any`, they're also useful without directly using that trait.

Not entirely sure how to summarize these together in one line though. Something about type erasure or dynamic typing or reflection maybe?

File: library/core/src/primitive_docs.rs [hunk 4]
```diff
@@ -731,12 +731,59 @@ mod prim_array {}
...
+/// ```
+///
+/// The mutable slice yields mutable references to the elements:
+///
+/// ```
+/// let mut scores = [7, 8, 9];
+/// for score in &mut scores[..] {
+///     *score += 1;
+/// }
+/// ```
```
Comment (m-ou-se) [inline library/core/src/primitive_docs.rs:763]: Both of these examples work with and without `[..]`, but only the `mut` version uses that now. The first one now iterates over a reference to an array, not over the (unsized) slice.

Maybe these should coerce the type explicitly to a slice, and then iterate over that without any `[..]`?

```suggestion
/// ```
/// let numbers: &[i32] = &[0, 1, 2];
/// for n in numbers {
///     println!("{} is a number!", n);
/// }
/// ```
///
/// The mutable slice yields mutable references to the elements:
///
/// ```
/// let scores: &mut [i32] = &mut [7, 8, 9];
/// for score in scores {
///     *score += 1;
/// }
/// ```
```

**Review by m-ou-se — APPROVED** [review]
This is great. Thanks for doing this.

I have a few small comments, but all of them are basically comments about the old documentation that you moved, so not really comments on your changes. If you feel like addressing them, go ahead. But `r=me` either way. :)

File: library/core/src/primitive_docs.rs (hunk 2 above)
Comment (camsteffen) [reply library/core/src/primitive_docs.rs:786]: I could just undo this change. Also is "slices" a bit over-technical or confusing? I'm not how sure "slice" applies to a `const &'static str` for example. How about "The primitive string type"?

File: library/core/src/any.rs (hunk 3 above)
Comment (camsteffen) [reply library/core/src/any.rs:2]: How about

> Utilities for dynamic typing or working with any type.

supposing `Any` and `TypeId` falls under "dynamic typing" and `type_name` is a util that works with any type.

**Comment by JohnCSimon** [comment]
ping from triage:
@camsteffen 
can you address the comments from m-ou-se?

FYI: when a PR is ready for review, post a message containing 
`@rustbot ready` to switch the PR to `S-waiting-on-review` so the PR appears in the reviewer's backlog.

**Comment by camsteffen** [comment]
My bad. Waiting for feedback.

@rustbot ready

File: library/core/src/primitive_docs.rs (hunk 2 above)
Comment (m-ou-se) [reply library/core/src/primitive_docs.rs:786]: "Slices" might be important to not cause confusion with the `String` type. 

> I'm not how sure "slice" applies to a `const &'static str` for example.

That's still a pointer+size to a 'slice' of the program (where the string is stored). But I suppose the main point is that if you take any substring, you still have the same type.

File: library/core/src/any.rs (hunk 3 above)
Comment (m-ou-se) [reply library/core/src/any.rs:2]: Yeah that could work. Or maybe "Utilities related to type erasure."? Either way is fine, as long as it doesn't imply it's only the `Any` type.

File: library/core/src/primitive_docs.rs (hunk 1 above)
Comment (camsteffen) [reply library/core/src/primitive_docs.rs:804]:
```suggestion
/// Here we have declared a string slice initialized with a string literal.
```

File: library/core/src/primitive_docs.rs (hunk 2 above)
Comment (camsteffen) [reply library/core/src/primitive_docs.rs:786]: I'll just revert this.

File: library/core/src/any.rs (hunk 3 above)
Comment (camsteffen) [reply library/core/src/any.rs:2]: Currently landed on "Utilities for dynamic typing or type reflection." I feel that "erasure" is a sub-concept of "dynamic typing", and `TypeId` and `type_name` are both utilities for "reflection", so that word is good to include.

**Comment by camsteffen** [comment]
@rustbot ready

Rebased and addressed comments.

**Comment by camsteffen** [comment]
Threw in a fix for variable capturing in `tidy_error!`.

**Comment by camsteffen** [comment]
Rebased. I'm not sure if I'm conflicting with the intent of #97434 which seems to make `std::error` into a page about general error handling best practices.

**Comment by JohnCSimon** [comment]
ping from triage: 
@camsteffen this has a merge conflict

**Comment by camsteffen** [comment]
Rebased. I decided to bulldoze the header line added to `core::any` in #91970 in order to keep it short.

File: library/alloc/src/slice.rs [hunk 5]
```diff
@@ -1,82 +1,12 @@
-//! A dynamically-sized view into a contiguous sequence, `[T]`.
+//! Utilities for the slice primitive type.
 //!
 //! *[See also the slice primitive type](slice).*
 //!
-//! Slices are a view into a block of memory represented as a pointer and a
-//! length.
+//! Most of the structs in this module are iterator types which can only be created
+//! using a certain function. For example, `slice.iter()` yields and [`Iter`].
```
Comment (Mark-Simulacrum) [inline library/alloc/src/slice.rs:6]:
```suggestion
//! using a certain function. For example, `slice.iter()` yields an [`Iter`].
```

File: library/core/src/primitive_docs.rs [hunk 6]
```diff
@@ -801,11 +801,57 @@ mod prim_array {}
 /// assert_eq!(2 * pointer_size, std::mem::size_of::<Box<[u8]>>());
 /// assert_eq!(2 * pointer_size, std::mem::size_of::<Rc<[u8]>>());
 /// ```
+///
+/// ## Trait Implementations
+///
+/// There are several implementations of common traits for slices. Some examples
+/// include:
+///
+/// * [`Clone`]
```
Comment (Mark-Simulacrum) [inline library/core/src/primitive_docs.rs:810]: Clone isn't implemented on `[T]` or `&[T]`  -- say more?

File: library/core/src/primitive_docs.rs [hunk 7]
```diff
@@ -801,11 +801,57 @@ mod prim_array {}
...
+/// * [`Eq`], [`Ord`] - for slices whose element type are [`Eq`] or [`Ord`].
+/// * [`Hash`] - for slices whose element type is [`Hash`].
```
Comment (Mark-Simulacrum) [inline library/core/src/primitive_docs.rs:812]: It feels a little odd that Eq/Ord are combined into one item here but Hash is separate. I think it makes sense to keep all lines separate or all together; I'd personally prefer separate.

File: library/core/src/primitive_docs.rs (hunk 6 above)
Comment (camsteffen) [reply library/core/src/primitive_docs.rs:810]: Hmm good point. I just moved this content from `std::slice` docs. I'll remove this line.

File: library/core/src/primitive_docs.rs (hunk 7 above)
Comment (camsteffen) [reply library/core/src/primitive_docs.rs:812]: This section could just be a couple sentences:

```rust
/// Some traits are implemented for slices if the element type implements
/// that trait. This includes [`Eq`], [`Hash`] and [`Ord`].
```

**Comment by camsteffen** [comment]
Addressed @Mark-Simulacrum's comments.

@rustbot ready

File: src/tools/tidy/src/lib.rs [hunk 8]
```diff
@@ -28,7 +28,7 @@ macro_rules! t {
 macro_rules! tidy_error {
     ($bad:expr, $fmt:expr) => ({
         *$bad = true;
-        eprintln!("tidy error: {}", $fmt);
+        eprintln!("tidy error: {}", format_args!($fmt));
```
Comment (Mark-Simulacrum) [inline src/tools/tidy/src/lib.rs:31]: Looks like an unrelated change?

**Comment by Mark-Simulacrum** [comment]
r=me with commits squashed and the unrelated tidy change dropped.

File: src/tools/tidy/src/lib.rs (hunk 8 above)
Comment (camsteffen) [reply src/tools/tidy/src/lib.rs:31]: Right. Do you want me to separate this?

Comment (camsteffen) [reply src/tools/tidy/src/lib.rs:31]: Oh nvm saw your other comment
